site/guides/variant-constructor-pattern.md

```diff
@@ -0,0 +1,90 @@
+
+Figure below shows a `package catalog` on the left that has `abstract` packages and `deployable instances` on the right. A good pattern is to keep the abstract packages and instance packages in separate repos and typically `deployable instances` repo will be setup to auto deploy to a kubernetes cluster using tools such as `config-sync`, `flux`, `argocd`.
+
+![variant constructor pkg repo diagram](img/variant-constructor-pkg-repo-diagram.png)
+
+
+Resources in an `abstract` package have placeholder values that needs to be substituated with actual values to make them ready for deployment. 
+For example, name of the namespace resource below has `example` as a placeholder value. This is a part of the `abstract package`.
+
+```yaml
+apiVersion:v1
+kind: Namespace
+metadata:
+	name: example # <-- this is a placeholder value
+```
+
+
+A kpt package contains kubernetes resources. Whenever you are creating a variant of the package, first step is to ensure unique identity of the resources in that variant. For example, if the abstract package contained a `namespace` resource, then the variant package will have namespace resource corresponding to that variant.
+
+In a kubernetes cluster, resources are identified by their group, version, kind, namespace and name (aka GVKNN). If resource is cluster scoped, then the `metadata.name` uniquely identifies the resource in a cluster. If the resource is namespace scoped, then (namespace, name) together identifies the resource uniquely.
+
+If you look at the steps involved in variant construction, it involves:
+1. set of functions that can help in updating unique identity of the resources
+2. Input to the functions for step 1
+
+[kpt-function-catalog](https://catalog.kpt.dev) provides two function that helps with customizing the identify of the resources:
+1. [set-namespace](https://catalog.kpt.dev/set-namespace/v0.3/): sets the namespace for all resources in the package.
+2. [ensure-name-substring](https://catalog.kpt.dev/ensure-name-substring/v0.2/): sets the name of the resources in the package.
+
+Now let's talk about the input to the functions. In most cases, variant's name (deployable package instance name) itself can be used to derive unique identity of the resources in the variant. For example, if I create a variant of `microservice` package, I will name the deployable package instance to `user-service` or `order-service`. So if the package's name is available to the functions, then they can use it to customize the name/namespace of the resources. So, starting with `kpt v1.0.0-beta.14+` , kpt makes `package's name` available in a local `ConfigMap` at a well-known path `pkg-context.yaml` that contains the package instance name and is available to functions during `kpt fn render` or `kpt fn eval`.
```

Comment:
Is it pkg-context.yaml or package-context.yaml?

Reply:
My bad. it is `package-context.yaml`. Fixed it in all places.